src/metabase/query_processor/util.clj

```diff
@@ -227,3 +229,16 @@
   "Returns `true` if query is an internal query."
   [{query-type :type} :- ::qp.schema/qp]
   (= :internal (keyword query-type)))
+
+(defn field->base-type
+  "Return base-type for a `field` clause. Field is expected to appear in the top level of the `_legacy-query`,
+  ie. not in its source query or source queries of its joins."
+  [{source-metadatas :source-metadata :as _legacy-query} [_tag id-or-name opts :as _field]]
+  (or (:base-type opts)
+      (when (integer? id-or-name)
+        (:base-type (lib.metadata/field (qp.store/metadata-provider) id-or-name)))
+      (when (string? id-or-name)
+        (some (fn [{:keys [name base_type] :as _source-metadata}]
+                (when (= name id-or-name)
+                  base_type))
+              source-metadatas))))
```

Comment:
Again I wonder what it means looking for the type in the source metadata. That doesn't seem to account for expressions, for example.

Reply:
Returning to this PR I've realized that base type is mandatory for string named fields. Hence this function is redundant in this effort. I will modify the code shortly.

_I believe_ that annotate already does that, or not? I assume the `metabase.query-processor.middleware.annotate/col-info-for-expression` is used for that.